<a href="https://colab.research.google.com/github/pouria1380-creator/Deep_learning_ne/blob/main/image_classification_cnn_torch.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Build a CNN model from scratch using PyTorch for image classification, train it on a real dataset, and evaluate its performance.

## Import libraries

### Subtask:
Import the necessary PyTorch libraries (torch, torch.nn, torch.optim, torchvision).


**Reasoning**:
Import the necessary PyTorch libraries as requested by the subtask.



In [14]:
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision

## Define the cnn model

### Subtask:
Create a class that inherits from `torch.nn.Module` and define the convolutional layers, pooling layers, and fully connected layers for the CNN.


**Reasoning**:
Define the CNN class with convolutional, pooling, and fully connected layers, and the forward pass.



In [15]:
class CNN(nn.Module):
    def __init__(self, num_classes=10):
        super(CNN, self).__init__()
        self.conv_layers = nn.Sequential(
            nn.Conv2d(3, 32, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2),
            nn.Conv2d(32, 64, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2),
            nn.Conv2d(64, 128, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2)
        )
        # Calculate the input size for the first fully connected layer
        # Assuming input image size is 32x32 (like CIFAR-10)
        # After three max pooling layers with kernel_size 2 and stride 2,
        # the spatial dimensions will be reduced by 2^3 = 8 times.
        # So, 32x32 becomes 4x4. The number of channels is 128.
        self.fc_layers = nn.Sequential(
            nn.Linear(128 * 4 * 4, 512),
            nn.ReLU(),
            nn.Linear(512, num_classes)
        )

    def forward(self, x):
        x = self.conv_layers(x)
        x = torch.flatten(x, 1) # Flatten the output for the fully connected layers
        x = self.fc_layers(x)
        return x

print("CNN class defined.")

CNN class defined.


## Load and preprocess data

### Subtask:
Load an image dataset (e.g., CIFAR-10 or MNIST) using `torchvision.datasets` and define data transformations.


**Reasoning**:
Define data transformations and load the CIFAR-10 dataset using torchvision.



In [16]:
import torchvision.transforms as transforms

# Define transformations
# Convert PIL Image to Tensor and normalize
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)) # Normalize CIFAR-10 images
])

# Load CIFAR-10 training dataset
trainset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)

# Load CIFAR-10 test dataset
testset = torchvision.datasets.CIFAR10(root='./data', train=False, download=True, transform=transform)

print("CIFAR-10 dataset loaded and transformations defined.")

100%|██████████| 170M/170M [00:11<00:00, 15.5MB/s]


CIFAR-10 dataset loaded and transformations defined.


## Create dataloaders

### Subtask:
Create `DataLoader` objects for training and testing data to handle batching and shuffling.


**Reasoning**:
Create DataLoader objects for the training and testing datasets using the specified batch size and shuffling requirements.



In [17]:
from torch.utils.data import DataLoader

# Create DataLoader for training dataset
trainloader = DataLoader(trainset, batch_size=64, shuffle=True)

# Create DataLoader for test dataset
testloader = DataLoader(testset, batch_size=64, shuffle=False)

print("DataLoaders created for training and testing.")

DataLoaders created for training and testing.


## Define loss function and optimizer

### Subtask:
Choose an appropriate loss function (e.g., CrossEntropyLoss for classification) and an optimizer (e.g., Adam or SGD).


**Reasoning**:
Instantiate the CrossEntropyLoss as the loss function and the Adam optimizer.



In [18]:
# Instantiate the model (assuming the CNN class is already defined from a previous step)
model = CNN(num_classes=10) # Assuming 10 classes for CIFAR-10

# Define the loss function
criterion = nn.CrossEntropyLoss()

# Define the optimizer
optimizer = optim.Adam(model.parameters(), lr=0.001)

print("Loss function and optimizer defined.")

Loss function and optimizer defined.


## Train the model

### Subtask:
Implement the training loop to iterate over epochs, calculate loss, perform backpropagation, and update weights.


**Reasoning**:
Implement the training loop as described in the instructions, iterating through epochs and batches to perform forward and backward passes and update model weights.



In [19]:
num_epochs = 10

print("Starting training...")
for epoch in range(num_epochs):
    model.train()  # Set the model to training mode
    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        # get the inputs; data is a list of [inputs, labels]
        inputs, labels = data

        # zero the parameter gradients
        optimizer.zero_grad()

        # forward + backward + optimize
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        # print statistics
        running_loss += loss.item()
        if i % 100 == 99:    # print every 100 mini-batches
            print(f'Epoch [{epoch + 1}/{num_epochs}], Batch [{i + 1}/{len(trainloader)}], Loss: {running_loss / 100:.4f}')
            running_loss = 0.0

print("Finished Training.")

Starting training...
Epoch [1/10], Batch [100/782], Loss: 1.9232
Epoch [1/10], Batch [200/782], Loss: 1.5507
Epoch [1/10], Batch [300/782], Loss: 1.4308
Epoch [1/10], Batch [400/782], Loss: 1.3296
Epoch [1/10], Batch [500/782], Loss: 1.2729
Epoch [1/10], Batch [600/782], Loss: 1.1993
Epoch [1/10], Batch [700/782], Loss: 1.1095
Epoch [2/10], Batch [100/782], Loss: 0.9903
Epoch [2/10], Batch [200/782], Loss: 0.9528
Epoch [2/10], Batch [300/782], Loss: 0.9716
Epoch [2/10], Batch [400/782], Loss: 0.9166
Epoch [2/10], Batch [500/782], Loss: 0.9057
Epoch [2/10], Batch [600/782], Loss: 0.8815
Epoch [2/10], Batch [700/782], Loss: 0.8538
Epoch [3/10], Batch [100/782], Loss: 0.7480
Epoch [3/10], Batch [200/782], Loss: 0.7224
Epoch [3/10], Batch [300/782], Loss: 0.7296
Epoch [3/10], Batch [400/782], Loss: 0.7095
Epoch [3/10], Batch [500/782], Loss: 0.7183
Epoch [3/10], Batch [600/782], Loss: 0.7037
Epoch [3/10], Batch [700/782], Loss: 0.7344
Epoch [4/10], Batch [100/782], Loss: 0.5937
Epoch [4/10

## Evaluate the model

### Subtask:
Implement the evaluation loop to calculate the model's performance on the test set (e.g., accuracy).


**Reasoning**:
Implement the evaluation loop to calculate the model's performance on the test set, including setting the model to evaluation mode, disabling gradient calculation, iterating through the test data, making predictions, and calculating accuracy.



In [20]:
# Implement the evaluation loop
print("\nStarting evaluation...")
model.eval()  # Set the model to evaluation mode
correct_predictions = 0
total_samples = 0

with torch.no_grad():  # Disable gradient calculation
    for data in testloader:
        images, labels = data
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1)
        total_samples += labels.size(0)
        correct_predictions += (predicted == labels).sum().item()

accuracy = 100 * correct_predictions / total_samples
print(f'Accuracy of the network on the 10000 test images: {accuracy:.2f} %')

print("Finished Evaluation.")


Starting evaluation...
Accuracy of the network on the 10000 test images: 74.93 %
Finished Evaluation.


## Summary:

### Data Analysis Key Findings

*   A Convolutional Neural Network (CNN) model was successfully defined using PyTorch, including convolutional, pooling, and fully connected layers.
*   The CIFAR-10 dataset was loaded and preprocessed with normalization.
*   DataLoaders were created for both training and testing datasets with a batch size of 64.
*   The CrossEntropyLoss was selected as the loss function, and the Adam optimizer with a learning rate of 0.001 was used.
*   The model was trained for 10 epochs, with the training loss decreasing over time.
*   The final trained model achieved an accuracy of 74.93% on the 10000 test images.

### Insights or Next Steps

*   The model's performance is reasonable but could potentially be improved by hyperparameter tuning (learning rate, batch size, number of epochs) or architectural changes (adding more layers, using different kernel sizes).
*   Further analysis could involve examining misclassified images to understand the model's weaknesses or implementing techniques like data augmentation to improve generalization.
